vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,468 @@
+    short node_track_num(RRNodeId node) const;
+    short node_class_num(RRNodeId node) const;
+
+    short node_cost_index(RRNodeId node) const;
+    e_direction node_direction(RRNodeId node) const;
+    e_side node_side(RRNodeId node) const;
+    float node_R(RRNodeId node) const;
+    float node_C(RRNodeId node) const;
+    RRSegmentId node_segment(RRNodeId node) const; /* get the segment id of a rr_node */
+
+    short node_num_configurable_in_edges(RRNodeId node) const;      /* get the number of configurable input edges of a node */
+    short node_num_non_configurable_in_edges(RRNodeId node) const;  /* get the number of non-configurable input edges of a node */
+    short node_num_configurable_out_edges(RRNodeId node) const;     /* get the number of configurable output edges of a node */
+    short node_num_non_configurable_out_edges(RRNodeId node) const; /* get the number of non-configurable output edges of a node */
+
+    /* Get a node-to-node switch */
+    std::vector<RRSwitchId> find_switches(RRNodeId from_node, RRNodeId to_node) const;
+
+    /* Get the range (list) of edges related to a given node */
+    edge_range node_configurable_in_edges(RRNodeId node) const;
+    edge_range node_non_configurable_in_edges(RRNodeId node) const;
+    edge_range node_configurable_out_edges(RRNodeId node) const;
+    edge_range node_non_configurable_out_edges(RRNodeId node) const;
+
+    edge_range node_out_edges(RRNodeId node) const;
+    edge_range node_in_edges(RRNodeId node) const;
+
+    //Edge attributes
+    size_t edge_index(RREdgeId edge) const; /* TODO: deprecate this accessor as outside functions should use RREdgeId */
+    RRNodeId edge_src_node(RREdgeId edge) const;
+    RRNodeId edge_sink_node(RREdgeId edge) const;
+    RRSwitchId edge_switch(RREdgeId edge) const;
+    bool edge_is_configurable(RREdgeId edge) const;
+    bool edge_is_non_configurable(RREdgeId edge) const;
+
+    /* Switch Info */
+    size_t switch_index(RRSwitchId switch_id) const; /* TODO: deprecate this accessor as outside functions should use RRSwitchId */
+    const t_rr_switch_inf& get_switch(RRSwitchId switch_id) const;
+
+    /* Segment Info */
+    size_t segment_index(RRSegmentId segment_id) const; /* TODO: deprecate this accessor as outside functions should use RRSegmentId */
+    const t_segment_inf& get_segment(RRSegmentId segment_id) const;
+
+    //Utilities
+    std::vector<RREdgeId> find_edges(RRNodeId src_node, RRNodeId sink_node) const;
+    RRNodeId find_node(short x, short y, t_rr_type type, int ptc, e_side side = NUM_SIDES) const;
+    short chan_num_tracks(short x, short y, t_rr_type type) const;
+
+    bool is_dirty() const;
+
+  public:                                 /* Echos */
+    void print_node(RRNodeId node) const; /* Print the detailed information of a node */
+  public:                                 /* Checkers */
+    /* Node-level checking */
+    bool check_node_segment(RRNodeId node) const;
```

Comment:
There has already been some discussion of these elsewhere. But I think only the top-level check or validation routine should be public. The others should be private as they are implementation details.